[Merged by Bors] - refactor(algebra/lie/basic): unbundle the action in `lie_module`

ocfnash opened the pull request:
---

1. The motivation for this work was a reminder from @jcommelin in [this Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/linear_action). The sought-for benefit is that we now have a notation for the action of a Lie algebra on its module. This is `⁅x, m⁆` for a Lie algebra element `x` acting on `m` in some module.

2. At first we thought we would instead use `has_scalar` so that the notation would be `x • m` rather than `⁅x, m⁆` but I worried that this would cause problems. This is because to support `lie_ideal`s we would have to define `has_scalar L L` which, in combination with `lie_algebra.of_associative_algebra`, would mean that commutative rings would carry two mathematically different scalar actions. The other one (used for associative `ideal`s) is: `example {R : Type*} [comm_ring R] : has_scalar R R := algebra.to_has_scalar`. For what it's worth I think there's [a chance we could get away with this](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/multiple.20solutions.20for.20typeclass.20search) but I think it's a bad idea to try.

3. The choice to use `has_bracket` to represent both the product in a Lie algebra, as well as the action of a Lie algebra on a module, together with the fact that Lie algebras are modules over themselves means that we need to be careful not to have to mathematically equal, but non-defeq brackets on a Lie algebras. The changes here achieve this but it was more challenging to organise everything so that this is true than I had expected.

4. **Many** apologies for the size of the changes proposed here, and for the multiple iterations.

Commit:
Allow heterogenous has_bracket

Commit:
Unbundle action in lie_module

Commit:
Remove simp attribute from `lie_smul` and `smul_lie`

These are now special cases of `lie_module_lie_smul` and
`lie_moduel_smul_lie`. As this latter pair carry the simp attribute,
the former should not since their LHS will not be in simp normal form.

Commit:
Add comment for `has_bracket`

Commit:
Fix derivations

Commit:
Simplify `lie_submodule_lie_module`

Commit:
Remove `lie_ring` test for `transport` tactic.

Unfortunately the new definition is out of reach of the tactic.

ocfnash commented on src/algebra/lie/basic.lean:
```diff
@@ -544,64 +549,139 @@ end lie_algebra
 algebra on this module, such that the Lie bracket acts as the commutator of endomorphisms. -/
-class lie_module extends linear_action R L M :=
-(lie_act : ∀ (l l' : L) (m : M), act ⁅l, l'⁆ m = act l (act l' m) - act l' (act l m))
+class lie_module (α : has_bracket L M) :=
```
Maybe worth noting that we could unbundle / break this down further as:
```lean
class lie_ring_module (L M : Type*) (α : has_bracket L M) [lie_ring L] [add_comm_group M] :=
(add_lie : ∀ (x y : L) (m : M), ⁅x + y, m⁆ = ⁅x, m⁆ + ⁅y, m⁆)
(lie_add : ∀ (x : L) (m n : M), ⁅x, m + n⁆ = ⁅x, m⁆ + ⁅x, n⁆)
(lie_lie : ∀ (x y : L) (m : M), ⁅⁅x, y⁆, m⁆ = ⁅x, ⁅y, m⁆⁆ - ⁅y, ⁅x, m⁆⁆)

class lie_algebra_module (R L M : Type*) (α : has_bracket L M)
  [comm_ring R] [lie_ring L] [lie_algebra R L] [add_comm_group M] [module R M] [lie_ring_module L M α] :=
(smul_lie : ∀ (t : R) (x : L) (m : M), ⁅t • x, m⁆ = t • ⁅x, m⁆)
(lie_smul : ∀ (t : R) (x : L) (m : M), ⁅x, t • m⁆ = t • ⁅x, m⁆)
```
to emphasise the analogy with the associative case:
```lean
class algebra_representation (R A M : Type*)
  [comm_ring R] [ring A] [algebra R A] [add_comm_group M] [module R M] [module A M]
  extends is_scalar_tower R A M
```

ocfnash replied:
Looking over this again now I think I might take another pass this weekend and see if I can carry out this further unbundling.

jcommelin commented on src/algebra/lie/basic.lean:
```diff
@@ -544,64 +549,139 @@ end lie_algebra
+  ⁅x + y, m⁆ = ⁅x, m⁆ + ⁅y, m⁆ :=
+lie_module.add_lie R x y m
+
+@[simp] lemma lie_module_lie_add
```
Would it be possible to call these simply `lie_add`, and also use them for brackets of `x y z : L`? Or does that have downsides?

ocfnash replied:
I agree this is desirable and will attempt to unify as much as possible when I return to this tomorrow.

ocfnash replied:
This is now done! The original `lie_add` etc. lemmas that were specific to Lie algebras / rings have been deleted and replaced with their more general counterparts which apply to Lie (ring) modules. I verified by hand that each of the lemmas removed could be proved `by simp` with the new lemmas in place.

jcommelin reviewed (commented):
Thanks for doing this. Looking pretty good. And the further unbundling that you suggest also sounds good to me.

Commit:
Further unbundling: lie_ring_module

This was trickier to get right than it might appear, especially
to organise in such a way that `lie_ring_self_module` does not introduce
a has_bracket instance that is not defeq equal to the original. This
complication does not arise in the associative case since there we have
has_scalar and has_mul, whereas here we are using just has_bracket.

The code still needs quite a bit of work but this is definite progress.

jcommelin commented on src/algebra/lie/basic.lean:
```diff
@@ -120,6 +125,12 @@ begin
+lemma jacobi (x y z : L) : ⁅x, ⁅y, z⁆⁆ + ⁅y, ⁅z, x⁆⁆ + ⁅z, ⁅x, y⁆⁆ = 0 :=
+begin
+  have h : ⁅z,⁅y,x⁆⁆ = -⁅z,⁅x,y⁆⁆, { rw ← lie_neg, congr, rw lie_skew, },
+  rw [←lie_skew, lie_ring.lie_lie, h], abel,
```
Do you need the `lie_ring.` prefix? Can't we have a global `lie_lie`?

ocfnash replied:
Done!

jcommelin commented on src/algebra/lie/basic.lean:
```diff
@@ -148,13 +159,13 @@ identity. Forgetting the scalar multiplication, every Lie algebra is a Lie ring.
 (lie_smul : ∀ (t : R) (x y : L), ⁅x, t • y⁆ = t • ⁅x, y⁆)
 
-@[simp] lemma lie_smul  (R : Type u) (L : Type v) [comm_ring R] [lie_ring L] [lie_algebra R L]
+lemma lie_smul (R : Type u) (L : Type v) [comm_ring R] [lie_ring L] [lie_algebra R L]
```
Maybe add a comment explaining why these aren't simp? (Future me would probably try to make them simp if I was browsing through this file.)

ocfnash replied:
These lemmas no longer exist so the problem is gone!

jcommelin commented on src/algebra/lie/basic.lean:
```diff
@@ -544,90 +540,145 @@ end lie_algebra
-
-protected lemma of_endo_map_action (α : L →ₗ⁅R⁆ module.End R M) (x : L) (m : M) :
-  @linear_action.act R _ _ _ _ _ _ _ (linear_action.of_endo_map R L M α) x m = α x m := rfl
+class lie_module [lie_ring_module L M] :=
```
Why do you assume `lie_ring_module` as argument, instead of extending it?

ocfnash replied:
Because `lie_module` has more carrier types (`R`, `L`, `M`) than `lie_ring_module` (`L`, `M`) so we would have the recursive typeclass search problem if I extended: I think we'd have loops where the algorithm spawns metavariables for `R` recursively when trying to find a `lie_ring_module` instance.

I.e., the same reason e.g., a regular `module` cannot extend `add_comm_group` but has to have it as an unbundled argument.

Commit:
Further tidy up

Commit:
Vanity golfing

ocfnash commented on src/algebra/lie/basic.lean:
```diff
@@ -42,119 +41,96 @@ lie bracket, ring commutator, jacobi identity, lie ring, lie algebra
+(lie_add : ∀ (x y z : L), ⁅x, y + z⁆ = ⁅x, y⁆ + ⁅x, z⁆)
 (lie_self : ∀ (x : L), ⁅x, x⁆ = 0)
-(jacobi : ∀ (x y z : L), ⁅x, ⁅y, z⁆⁆ + ⁅y, ⁅z, x⁆⁆ + ⁅z, ⁅x, y⁆⁆ = 0)
+(lie_lie : ∀ (x y z : L), ⁅⁅x, y⁆, z⁆ = ⁅x, ⁅y, z⁆⁆ - ⁅y, ⁅x, z⁆⁆)
```
Looking now I wonder if the following alternative to `lie_lie` would be preferable:
```lean
(leibniz_lie : ∀ (x y z : L),  ⁅x, ⁅y, z⁆⁆ = ⁅⁅x, y⁆, z⁆ + ⁅y, ⁅x, z⁆⁆)
```
and similarly for `lie_ring_module`.

jcommelin replied:
In the end, it doesn't really matter, right? We will need all variations.
We might want to provide several different constructors...

ocfnash replied:
Yeah, it doesn't really of course.

I was just thinking it's slightly nicer not to require subtraction when it's not necessary. And who knows, maybe some day we'll enter the exotic / somewhat hilarious world of Lie rings over monoids that are not necessarily groups?

jcommelin replied:
Ok, let's go with `leibniz_lie`.

ocfnash replied:
Done!

Commit:
Change definition of lie_ring, lie_ring_module to avoid using subtraction

jcommelin reviewed (commented):
LGTM, please merge once CI is happy:

bors d+

eric-wieser commented on src/algebra/lie/basic.lean:
```diff
@@ -42,119 +41,99 @@ lie bracket, ring commutator, jacobi identity, lie ring, lie algebra
-class has_bracket (L : Type v) := (bracket : L → L → L)
+/-- The has_bracket class has two intended uses:
+
+  1. for the product `⁅x, y⁆` of two elements `x`, `y` in a Lie algebra
+     (analogous to `has_mul` in the associative setting),
+
+  2. for the action `⁅x, m⁆` of an element `x` of a Lie algebra on an element `m` in one of its
+     modules (analogous to `has_scalar` in the associative setting). -/
```
It strikes me as slightly odd to have one notation for the lie setting but two notations for the associative setting - for instance, should the signature of `has_mul` be changed to `A → B → B` so that it can subsume `has_scalar`?

ocfnash replied:
I too would prefer a one-to-one correspondence but I do not think we should try to subsume `has_mul` into `has_scalar`. I don't have great reasons though, really just:

- killing `has_mul` would have the disadvantage of needing extra type ascriptions in various places,
- though (I hope!) I made it look easy, it was actually fairly tricky to set things up in such a way that we never end up with two non-defeq `has_bracket` instances, and I bet it would be a hard refactor to achieve this in the much larger library of associative algebra.
